### PR TITLE
Fix release packaging, update LOOT API

### PR DESCRIPTION
**Pull request by WrinklyNinja**
The setup.py generated when packaging for release isn't necessarily recognised as executable by the OS, which was causing me problems, so run it using the Python executable.

I've also lumped in an update to the latest release of the LOOT API module, with it WB is able to start for me. Relevant to #388.

### CODE REVIEW

**Comment by Utumno** [comment]
Thanks will test ASAP - or @D4id4los if you can give that a whirl

**Comment by Utumno** [comment]
Unfortunately, my build still crashes - would it be useful to you to produce some kind of output from visual studio?

I build it alright but then when I try to run it crashes (windows dialog, stop the program)

**Comment by Utumno** [comment]
I posted my build for forum folks to test - as always in https://www.dropbox.com/sh/7b5ily482re0khs/AAD3vAWMVJNIpuS30tsdVte5a?dl=0

**Comment by WrinklyNinja** [comment]
Hmm, that standalone build works for me. I wonder if there's some runtime I've got as part of the developer tools that you don't? I'll try searching around for info on this.

**Comment by WrinklyNinja** [comment]
@Utumno can you check if you're able to run [this build](https://bintray.com/wrinklyninja/loot/download_file?file_path=loot_0.12.5-15-g0388f4f_dev.7z) of LOOT? It uses the same API DLL, if you can run that it narrows the problem down to the Python wrapper.

**Comment by Utumno** [comment]
@WrinklyNinja: I am able to run it (unpacked then run LOOT.exe). I had a crash window pop up along with the AVG scan - but then it launched normally. Bash still won't run - here is the process monitor log if you can make anything of it:

[Logfile process monitor WB 307.201804202117 loot 13 crash.zip](https://github.com/wrye-bash/wrye-bash/files/1934902/Logfile.process.monitor.WB.307.201804202117.loot.13.crash.zip)

**Comment by WrinklyNinja** [comment]
I couldn't spot anything useful in that log. Could it be AVG interfering with Bash's loot_api.dll?

**Comment by WrinklyNinja** [comment]
Have you got this [MSVC 2015 Update 3 redistributable](https://www.microsoft.com/en-us/download/details.aspx?id=53587) installed? It's not the one that LOOT's releases link to, but I noticed that pybind11 was specific about requiring 2015u3 and not just 2015, so maybe there's additional functionality there?

**Comment by Utumno** [comment]
Yep that fixed it (had to uninstall previous one and install x86) - the same standalone now starts !! :confetti_ball: :fireworks:

Time to update the install_loot script I guess - but we should uninstall the other if it exists!

**Comment by Utumno** [comment]
Just a minute I started the dev one

**Comment by WrinklyNinja** [comment]
Interesting, I should probably update LOOT's links too, since everything's built using the same version of MSVC, so it must just be luck that the older 2015 redist works.

I wouldn't uninstall any existing 2015 redist though, something else the user has may also need it.

**Comment by Utumno** [comment]
Wait just a moment - I had forgotten to install the new loot api was using the opld one (crossing fingers)

**Comment by Utumno** [comment]
Nope :((((( Still crashes

**Comment by nallar** [comment]
From your process explorer log, it tried to load `api-ms-win-core-fibers-l1-1-1.DLL` but could not.

That could be related to the crash. Apparently that DLL is [included in windows 10](https://docs.microsoft.com/en-us/uwp/win32-and-com/win32-apis).

You would think if that was a required part of the 2015 CRT for windows 7 the CRT redistributable would install it. 😕 It could be a soft requirement and is not used if not found, so that wold not be the cause of the crash.

I don't think this will fix it, but you could try installing the 2017 x86 CRT from here as well: https://support.microsoft.com/en-gb/help/2977003/the-latest-supported-visual-c-downloads

Visual studio's [just in time debugger](https://msdn.microsoft.com/en-us/library/5hs4b7a6.aspx) will let you start debugging when a process crashes.

**Comment by Utumno** [comment]
@nallar - I have the same dll errors in the dev branch standalone that however does not crash. Will try installing it and see if this fixes it.
I have tried using the VS debugger but I can't make head or tails - I am presented with an assembly dump that I can't really grok

**Comment by nallar** [comment]
Debugging without symbols/source code is not fun. If you build the loot API on your machine in debug mode and use that DLL that would help.

Can you copy/screenshot the call stack and exception when it crashes?

As WrinklyNinja built the DLL, you could create a dump file of the crashed process and give it to him to look at as he will have the correct PDB. Might be easier.

https://windowsdebugging.wordpress.com/2017/06/16/creating-process-dump-file-from-visual-studio/

**Comment by WrinklyNinja** [comment]
> As WrinklyNinja built the DLL, you could create a dump file of the crashed process and give it to him to look at as he will have the correct PDB. Might be easier.

I don't have the PDB as the builds happen on AppVeyor, I don't publish local builds. I'm in the process of updating the AppVeyor builds to debug info and package PDBs though.

**Comment by WrinklyNinja** [comment]
@Utumno Try with [this build](https://bintray.com/wrinklyninja/loot/download_file?file_path=loot_api_python-3.1.0-2-g6748f6b_relwithdebinfo-build-win32.7z). The PDB included is for the Python wrapper, I'm not sure how well it'll work since that's generated code. If the PDB for the LOOT API DLL would be more useful, it's in [this archive](https://bintray.com/wrinklyninja/loot/download_file?file_path=loot_api-0.13.2-1-gc4fd29a_relwithdebinfo-build-win32.7z).

**Comment by Utumno** [comment]
I tried https://bintray.com/wrinklyninja/loot/download_file?file_path=loot_api-0.13.2-1-gc4fd29a_relwithdebinfo-build-win32.7z - dumped

loot_api.pyd
loot_api.dll
loot_api.pdb

to the Mopy folder and built the standalone. Here is what I see in VS:

loot_api_python-3 1 0-2-visual_studio

and the dump (thanks nallar):

Not sure I used it correctly @WrinklyNinja (also should I use the dll from `loot_api-0.13.2-1-gc4fd29a_relwithdebinfo-build-win32.7z` and replace the one I used ?)  - I guess this is from py2exe part of the code, this apparently needs to also be built with symbols but I don't have no clue how.

Can you build a Standalone @WrinklyNinja and test if it runs then link me to it to see if it's the WB build process or the environment that cause my crash ?

**Comment by Utumno** [comment]
New iteration - I dumped all the content from loot_api-0.13.2-1-gc4fd29a_relwithdebinfo-build-win32.7z to the mopy folder
loot_api_python-3 1 0-2-visual_studio-symbols_loaded

Loaded the symbols for `loot_api.dll` but still no source - it's call from the dll that crashes it seems but still no source

**Comment by WrinklyNinja** [comment]
> Can you build a Standalone @WrinklyNinja and test if it runs then link me to it to see if it's the WB build process or the environment that cause my crash ?

Here's a build I did with the DLL/wrapper I linked to earlier, it works for me: [Mopy.zip](https://github.com/wrye-bash/wrye-bash/files/1969816/Mopy.zip).

I've only ever debugged executables that I've built and at runtime, so I don't really know what to do, but the exception you've screenshotted above is coming from `Wrye Bash.exe`, so it's not even getting to the external LOOT API DLL, so its debug symbols probably aren't the ones you want. It might be something to do with the wrapper, but that's all generated code so I don't know how helpful having the symbols will be.

Just to check: you don't have any issues running the non-standalone Bash with the LOOT API?

**Comment by Utumno** [comment]
> Here's a build I did with the DLL/wrapper I linked to earlier, it works for me: Mopy.zip.

And fails for me same error - so it's something in the environment WB.exe is run ?

>  so I don't really know what to do, but the exception you've screenshotted above is coming from Wrye Bash.exe, so it's not even getting to the external LOOT API DLL, so its debug symbols probably aren't the ones you want

However I do get (sometimes) the "Source not available" "Frame not in module" when I place the 64 mb `loot_api.pdb` by the exe - I also included `loot_api.lib` - _maybe_ including the source for that lib would help ? I do see that the unhandled exceptions is thrown from WB exe, just maybe it this "Source not available" refers to the dll - which commit corresponds to that build ?

I did try to find info on building py2exe with debug options but had no luck.

To recap - same exe fails for me works for you and others, and previous exe build (dev branch code with the previous loot api) works for everybody. Any chance we figure out from the tree what change caused this?

**Comment by WrinklyNinja** [comment]
> And fails for me same error - so it's something in the environment WB.exe is run ?

Hmm, I guess so. Time to do more digging...

I don't think `loot_api.lib` would be of any use when debugging, it contains info useful when compile-time linking to the corresponding DLL. The `source not available` will be because the debugger can't find the LOOT API's source code, which makes sense because it doesn't ship with the DLL, but I guess that would only be useful for seeing where in the code something is going wrong, or stepping through code, but we're doing post-run analysis anyway.

> Any chance we figure out from the tree what change caused this?

I can bisect history to test out different changesets, as the exception is thrown from `Wrye Bash.exe` that makes me think it's coming from the LOOT API's Python wrapper, so I'll focus on its changes (which are thankfully much fewer).

**Comment by Utumno** [comment]
> I don't think loot_api.lib would be of any use when debugging, it contains info useful when compile-time linking to the corresponding DLL. The source not available will be because the debugger can't find the LOOT API's source code

I understand this and I am still not sure I can plug it in to VS but maybe worth a try (which commit does this dll correspond to?) - although as you say this is probably a call from (or to) the `pyd` and adding debug info to the py2exe built doesn't seem possible

Compiled mysteries

**Comment by WrinklyNinja** [comment]
@Utumno if you revert @D4id4los's [changes to mods_metadata.py](https://github.com/wrye-bash/wrye-bash/commit/6f99da8defbceefc2dbd92b6c24f8e4f1835684e) and build an exe using the [API module v2.0.0](https://github.com/loot/loot-api-python/releases/download/2.0.0/loot_api_python-2.0.0-0-gabbc089_master-win32.7z), does it still crash for you?

**Comment by Utumno** [comment]
Sorry it took so long @WrinklyNinja - I reverted the changes and built with the api you provided - I still get a crash but that's inside bash:

```
Traceback (most recent call last):
  File "bash\bash.pyo", line 227, in main
  File "bash\bash.pyo", line 393, in _main
  File "bash\basher\__init__.pyo", line 3981, in Init
  File "bash\basher\__init__.pyo", line 4018, in InitData
  File "bash\bosh\__init__.pyo", line 1956, in refresh
  File "bash\bosh\__init__.pyo", line 2117, in reloadBashTags
  File "bash\bosh\__init__.pyo", line 712, in reloadBashTags
  File "bash\bosh\mods_metadata.pyo", line 309, in getTagsInfoCache
RuntimeError: bad cast
```

but I guess this is ok

**Comment by WrinklyNinja** [comment]
OK, that RuntimeError will be coming from the LOOT API or its wrapper, but the important thing is that loading it works.

Try with the [2.1](https://bintray.com/wrinklyninja/loot/download_file?file_path=loot_api_python-2.0.0-1-g1e0dbe8_2.1-win32.7z) and [2.2](https://bintray.com/wrinklyninja/loot/download_file?file_path=loot_api_python-2.0.0-2-g212f899_2.2-win32.7z) branches, I couldn't think of better branch names, but 2.1 updates the LOOT API to 0.11.1, and 2.2 also updates pybind11.

**Comment by Utumno** [comment]
2.2 crashes - 2.1 fails with the `bad cast` - so it looks the change in pybind11 may be responsible.

Both display the same mktglist error:

```
C:\Dropbox\eclipse_workspaces\python\wrye-bash\scripts>mktaglist.py
Loaded the LOOT API v0.11.1 using wrapper version 2.0.0
[2018-05-11 23:45:01.926999] [0x00002cdc] [info]    Initialising load order data for game of type 0 at: "c:\users\mrd\appdata\local\temp\tmprjjumi"
[2018-05-11 23:45:01.928999] [0x00002cdc] [debug]   Loading file: "c:\users\mrd\appdata\local\temp\tmprjjumi\masterlist.yaml"
[2018-05-11 23:45:02.436028] [0x00002cdc] [trace]   Testing condition syntax: active("Oblivion.esm") and active("Nehrim.esm")
[2018-05-11 23:45:02.436028] [0x00002cdc] [trace]   Checking to see if the path ""Oblivion.esm"" is safe.
[2018-05-11 23:45:02.436028] [0x00002cdc] [trace]   Active check result: false
[2018-05-11 23:45:02.437029] [0x00002cdc] [trace]   Checking to see if the path ""Nehrim.esm"" is safe.
[2018-05-11 23:45:02.437029] [0x00002cdc] [trace]   Active check result: false
Traceback (most recent call last):
  File "C:\Dropbox\eclipse_workspaces\python\wrye-bash\scripts\mktaglist.py", line 65, in <module>
    lootDb.load_lists(masterlistPath)
RuntimeError: bad cast

C:\Dropbox\eclipse_workspaces\python\wrye-bash\scripts>mktaglist.py
Loaded the LOOT API v0.11.1 using wrapper version 2.0.0
[2018-05-11 23:45:28.078495] [0x00001308] [info]    Initialising load order data for game of type 0 at: "c:\users\mrd\appdata\local\temp\tmpptg0vn"
[2018-05-11 23:45:28.079495] [0x00001308] [debug]   Loading file: "c:\users\mrd\appdata\local\temp\tmpptg0vn\masterlist.yaml"
[2018-05-11 23:45:28.587524] [0x00001308] [trace]   Testing condition syntax: active("Oblivion.esm") and active("Nehrim.esm")
[2018-05-11 23:45:28.587524] [0x00001308] [trace]   Checking to see if the path ""Oblivion.esm"" is safe.
[2018-05-11 23:45:28.587524] [0x00001308] [trace]   Active check result: false
[2018-05-11 23:45:28.588524] [0x00001308] [trace]   Checking to see if the path ""Nehrim.esm"" is safe.
[2018-05-11 23:45:28.588524] [0x00001308] [trace]   Active check result: false
Traceback (most recent call last):
  File "C:\Dropbox\eclipse_workspaces\python\wrye-bash\scripts\mktaglist.py", line 65, in <module>
    lootDb.load_lists(masterlistPath)
RuntimeError: bad cast
```

The Traceback line that 2.1 crashes at is:

```python
                tags = lootDb.get_plugin_tags(modName.s)
```

**Comment by WrinklyNinja** [comment]
OK, [here's](https://bintray.com/wrinklyninja/loot/download_file?file_path=loot_api_python-3.1.0-4-gdaa3a49_relwithdebinfo-build-win32.7z) a build of the latest code but with pybind11 rolled back to the same version as in v2.0.0 of the wrapper.

**Comment by WrinklyNinja** [comment]
Well, it launches for me, but I don't see tags that are in the Skyrim SE taglist displayed in the UI. I checked with the Python version and same deal, and the last line in the BashBugDump.log is `mods_metadata.py  229 __init__: Using LOOT API version: 0.13.2`, so it's not immediately obvious why it's not working.

Now that we've got a build that will actually load though, I'll try to see why the integration isn't working, and go through and tidy things up for a proper release. Though I've probably only got a few hours till I'm AFK for a week.

On a related note, is it possible to launch the exe in debug mode like you can the Python version? I know you can enable debug logging once launched, but that doesn't catch any init issues.

**Comment by WrinklyNinja** [comment]
It looks like something is trying to get tags before the taglist has loaded. Here's an excerpt from my log after putting in a few `deprint()` statements:

```
mods_metadata.py  229 __init__: Using LOOT API version: 0.13.1
mods_metadata.py  265 refreshBashTags: Refreshing bash tags
mods_metadata.py  270 refreshBashTags: Masterlist exists at  C:\Users\Oliver\AppData\Local\LOOT\Skyrim Special Edition\masterlist.yaml
mods_metadata.py  281 refreshBashTags: Loading masterlist at  C:\Users\Oliver\AppData\Local\LOOT\Skyrim Special Edition\masterlist.yaml and userlist at C:\Users\Oliver\AppData\Local\LOOT\Skyrim Special Edition\userlist.yaml
mods_metadata.py  316 getTagsInfoCache: Getting tags for  Bashed Patch, 0.esp
mods_metadata.py  318 getTagsInfoCache: Got tags  set([])
mods_metadata.py  316 getTagsInfoCache: Getting tags for  Blank.esp
mods_metadata.py  318 getTagsInfoCache: Got tags  set([])
mods_metadata.py  316 getTagsInfoCache: Getting tags for  Update.esm
mods_metadata.py  318 getTagsInfoCache: Got tags  set([])
mods_metadata.py  316 getTagsInfoCache: Getting tags for  Skyrim.esm
mods_metadata.py  318 getTagsInfoCache: Got tags  set([])
mods_metadata.py  316 getTagsInfoCache: Getting tags for  Dawnguard.esm
mods_metadata.py  318 getTagsInfoCache: Got tags  set([])
mods_metadata.py  316 getTagsInfoCache: Getting tags for  ccqdrsse001-survivalmode.esl
mods_metadata.py  318 getTagsInfoCache: Got tags  set([])
mods_metadata.py  316 getTagsInfoCache: Getting tags for  HearthFires.esm
mods_metadata.py  318 getTagsInfoCache: Got tags  set([])
mods_metadata.py  316 getTagsInfoCache: Getting tags for  Dragonborn.esm
mods_metadata.py  318 getTagsInfoCache: Got tags  set([])
mods_metadata.py  265 refreshBashTags: Refreshing bash tags
mods_metadata.py  270 refreshBashTags: Masterlist exists at  C:\Users\Oliver\AppData\Local\LOOT\Skyrim Special Edition\masterlist.yaml
mods_metadata.py  297 refreshBashTags: Going to load taglist
mods_metadata.py  302 refreshBashTags: Loading taglist at  C:\Users\Oliver\Documents\Code\wrye-bash\Mopy\Bash Patches\Skyrim Special Edition\taglist.yaml
```

(The printed tag sets are `tags.added`.)

**Comment by Utumno** [comment]
Thanks @WrinklyNinja yes some cleanup is needed - loot_api_download updated also.

Info on pre wx booting should be in Wrye Bash.exe.log - see: https://github.com/wrye-bash/wrye-bash/wiki/%5Bgithub%5D-Reporting-a-bug#bash-log-files

This is a bit grey land those log files but no time for that now

Btw the dll/api sizes were much bigger - not sure if matters

**Comment by WrinklyNinja** [comment]
> Btw the dll/api sizes were much bigger - not sure if matters

That's due to them shipping with debug info.

**Comment by Utumno** [comment]
Ooops - will have to track `getTagsInfoCache` - don't spend too much brain power (and valuable keyboard time :P)

Be aware that probably the tags are filtered to only include supported ones (per game - track `allTags` global)

**Comment by Utumno** [comment]
> That's due to them shipping with debug info.

I guessed so - should we link to the debug binaries in release (loot_api script for instance ?)

**Comment by WrinklyNinja** [comment]
> I guessed so - should we link to the debug binaries in release (loot_api script for instance ?)

They're still release binaries, just with debug info. From what I've read the size increase is the only significant impact, performance should be equivalent, and being able to use the debug symbols may come in handy again.

**Comment by Utumno** [comment]
So in they go - " use the debug symbols" is quite another story - was not able to (we need the source too and somewhere where VS can find it)  nor have I a clue how to add debug info to the Wrye Bash exe.

But at some point we should revisit so we can really see the line that crashes in C++

**Comment by Utumno** [comment]
Don't forget to add a note on dll/pyd version numbers :P (say in loot_api.py ?) - maybe bump them to the same number?

**Comment by Utumno** [comment]
Will do ASAP and pack a standalone for people to test (along with fixes for a couple bugs reported in the forums) - have a nice AFK week!

**Comment by Utumno** [comment]
Hi @WrinklyNinja - so I keep:

```
+lootApiWrapperVersion = '3.1.1'
+lootApiWrapperRevision ='f97de90'
```

or do you have a newer one I need to try? I need to release before summer comes for good...

this one is working fine on my part - I build and run

**Comment by WrinklyNinja** [comment]
Hmm, I should probably release an update for the API v0.13.x, I'll get it done this weekend. Would you like a new PR, or should I update this one?

**Comment by Utumno** [comment]
Well actually the beta went out with that version (I was planning to delay this but it was published by alt3rn1ty - so much the better I'd say)

I would like a new pull with the new API yes (will close this one from the commit on dev when I get round to merging). This will be merged in 307 proper - I expect this to be the last beta
